File: domain/policies/top.py
"""Pure policies for top-by-popularity browse categories.

Add a new category by appending to ``TOP_CATEGORY_SPECS`` (and mapping it
in each provider's ``top()``). Validation, labels, and local status seeds
all derive from that registry.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from domain.errors import ValidationError


@dataclass(frozen=True, slots=True)
class TopCategorySpec:
    """Descriptor for one top-browse category."""

    key: str
    label: str
    local_status: Optional[str] = None


# Ordered registry — UI tabs follow this order.
TOP_CATEGORY_SPECS: tuple[TopCategorySpec, ...] = (
    TopCategorySpec(key="all", label="All"),
    TopCategorySpec(key="airing", label="Airing", local_status="AIRING"),
    TopCategorySpec(key="upcoming", label="Upcoming", local_status="UPCOMING"),
)

TOP_CATEGORIES: frozenset[str] = frozenset(spec.key for spec in TOP_CATEGORY_SPECS)

_TOP_LOOKUP: dict[str, TopCategorySpec] = {
    spec.key: spec for spec in TOP_CATEGORY_SPECS
}


def normalize_top_category(value: str) -> str:
    """Return a canonical top-category key or raise ``ValidationError``."""
    normalized = (value or "").strip().lower()
    if normalized not in _TOP_LOOKUP:
        allowed = ", ".join(spec.key for spec in TOP_CATEGORY_SPECS)
        raise ValidationError(f"Top category must be one of: {allowed}.")
    return normalized


def top_category_spec(value: str) -> TopCategorySpec:
    """Return the registry entry for a validated category key."""
    return _TOP_LOOKUP[normalize_top_category(value)]


def local_status_for(category: str) -> Optional[str]:
    """Local catalog status seed for a category, or ``None`` if provider-only."""
    return top_category_spec(category).local_status


def format_top_label(category: str) -> str:
    """Human label such as ``Airing`` or ``All``."""
    return top_category_spec(category).label

File: next-web/lib/top.ts
/** Top-by-popularity browse helpers for the Next.js library UI. */

export const TOP_CATEGORY_SPECS = [
  { key: "all", label: "All" },
  { key: "airing", label: "Airing" },
  { key: "upcoming", label: "Upcoming" },
] as const;

export type TopCategory = (typeof TOP_CATEGORY_SPECS)[number]["key"];

export const TOP_CATEGORIES = TOP_CATEGORY_SPECS.map((spec) => spec.key);

const TOP_LOOKUP = new Map(
  TOP_CATEGORY_SPECS.map((spec) => [spec.key, spec] as const),
);

export function isValidTopCategory(value: string): value is TopCategory {
  return TOP_LOOKUP.has(value.trim().toLowerCase() as TopCategory);
}

export function normalizeTopCategory(value: string): TopCategory | null {
  const normalized = value.trim().toLowerCase();
  return TOP_LOOKUP.has(normalized as TopCategory) ? (normalized as TopCategory) : null;
}

export function formatTopLabel(category: string): string {
  const normalized = normalizeTopCategory(category);
  if (!normalized) return "Top";
  return TOP_LOOKUP.get(normalized)?.label ?? "Top";
}

export function topBrowseUrl(
  category: string,
  params: { page?: number; size?: number } = {},
): string {
  const normalized = normalizeTopCategory(category);
  if (!normalized) return "/library/top";
  const parts = [`category=${encodeURIComponent(normalized)}`];
  const page = params.page ?? 1;
  if (page > 1) parts.push(`page=${page}`);
  const size = params.size;
  if (size != null && size !== 24) parts.push(`size=${size}`);
  return `/library/top?${parts.join("&")}`;
}

export function parseTopBrowseParams(category?: string): TopCategory | null {
  if (!category) return null;
  return normalizeTopCategory(category);
}

export function defaultTopCategory(): TopCategory {
  return "all";
}
